### PR TITLE
Update all blocks to V3

### LINKED ISSUE #48716: [Enhancement]: Update blocks API version to V3
WordPress 6.3 added support for Block API V3, blocks declared as V3 will be loaded in an iframe, anything not declared as V3 will cause the editor to load in regular style, not iframe, which causes inconsistent styles.

https://developer.wordpress.org/block-editor/reference-guides/block-api/block-api-versions/

The main issue is that if someone has WooCommerce installed, their editor will always load up without iframe, regardless if they're using our blocks or not in such pages.

Things to look out for styles inconsistiens and usage of global packages (jquery for example).

Monorepo has around 163 block.json file, with a single block being on V1 (All Products), 160 blocks on V2, and a single block on V3 (`product-collection-no-results`), and one block that I didn't find.

Not all blocks seems to have a block.json or actually use it, so this is going to be an inconsistent long journy.

### Expected Result

- Gutenberg is loaded via an iframe.
- All Blocks continue to function fine (styling and behavior).
- This snippet returns true `!!document.querySelector('iframe[name=editor-canvas]')`

### Current behavior

- Gutenberg is not loaded in iframe.
- This snippet is returing false `!!document.querySelector('iframe[name=editor-canvas]')`

### Caviat

Gutebnerg will only switch to iframe if all registered blocks are on V3, so we can't test for regressions unless everything is on V3.

However, if you install latest Gutenberg plugin and enable an FSE theme, Gutenberg will always load up in iframe, which means you can test gradually.

You can also ensure no blocks are on V2 via this code `wp.data.select('core/blocks').getBlockTypes().filter(block => block.apiVersion !== 3)`

### LINKED ISSUE #50415: In the editor, the Checkout block's step titles have vertical scroll bars
### Prerequisites

- [ ] I have carried out troubleshooting steps and I believe I have found a bug.
- [ ] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

In the editor, the Checkout block's step titles have vertical scroll bars even though the text is on one line.
image

### Expected behavior

For the vertical scrollbars not to appear

### Actual behavior

In the editor, the Checkout block's step titles have vertical scroll bars

### Steps to reproduce

1. Go to the Checkout block in the editor (site editor or page editor is fine)
2. Notice the step titles have vertical scroll bars.

### WordPress Environment

n/a

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by senadir**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR updates all WooCommerce blocks to API Version 3. By doing so, given that the store doesn't have any blocks by other plugins using version 1 or 2, the post editor will use iframes like the site editor, making blocks work consistently among editors.

Closes #48716 
Closes https://github.com/woocommerce/woocommerce/issues/50415

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### With only WooCommerce activated
1. Add a new post/page, see the canvas using an iframe.
2. Smoke tests all blocks provided by WooCommerce, ensuring they are working normally.
3. Save the post, go to the front end, and see all blocks work as expected.
4. Go to Site Editor, and smoke tests all blocks provided by WooCommerce, ensuring they are working normally in the editor and on the front end.

#### With a plugin/theme containing a v1 or v2 block.
1. Activate a plugin containing a block using a previous API version. You can install and activate this example plugin: https://github.com/dinhtungdu/just-another-v2-block.
2. Add a new post/page, see the canvas isn't using an iframe, but a div element.
3. Smoke tests all blocks provided by WooCommerce, ensuring they are working normally.
3. Save the post, go to the front end, and see all blocks work as expected.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Comment by dinhtungdu** [comment]
I smoke-tested all blocks maintained by Kirigami and Origami and they all work as expected.

**Comment by senadir** [comment]
Thank you @dinhtungdu can you or someone from your team confirm none of the broken tests are:
- Owned by you.
- Due to enabling iframes

**Comment by dinhtungdu** [comment]
@senadir it seems that I overlooked somethings when testing all blocks at the same time. The All Products block isn't rendered at all on the front end. I will go through all failed tests and share a list of affected block later.

**Comment by dinhtungdu** [comment]
I went through all the blocks again and failed tests. Only the All Products block and its variations have the render issue and need to be updated:

- Issue: The block isn’t rendered on the front end.
- Reason: The save function doesn’t use the ‘useBlockProps’ hook. With v1, the `.wp-block-woocommerce-all-products`  class is included in the saved content, but it isn’t with v2 or v3. The frontend renderer of the All Products block uses that class to render the block.
- Affected E2E tests:
    - all-products.block\_theme.spec.ts:18  › block can be inserted and it is rendered on the frontend
    - price-filter.block\_theme.spec.ts:180 › should show only products that match the filter
    - price-filter.block\_theme.spec.ts:163 › should show all products

  

Most failed E2E tests are because we use `page`  directly instead of `editor.canvas`, which is expected to fail with iframe:

- all-reviews.block\_theme.spec.ts:29 › block can be inserted and it sorts reviews by most recent by default
- attribute-filter.block\_theme.spec.ts:43 › should allow changing the block's title
- attribute-filter.block\_theme.spec.ts:54 › should allow changing the display style
- price-filter.block\_theme.spec.ts:47 › Filter by Price Block - editor side › should allow changing the block's title
- product-on-sale.block\_theme.spec.ts:12 › woocommerce/product-on-sale Block › can be inserted in Post Editor and it is visible on the frontend
- rating-filter.block\_theme.spec.ts:50 › Filter by Rating Block › should allow changing the display style
- rating-filter.block\_theme.spec.ts:39 › Filter by Rating Block › should allow changing the block's title
- reviews-by-category.block\_theme.spec.ts:29 › woocommerce/reviews-by-category Block › block can be inserted and it successfully
- reviews-by-product.block\_theme.spec.ts:29 › woocommerce/reviews-by-product Block › block can be inserted and it successfully
- stock-filter.block\_theme.spec.ts:51 › Filter by Stock Block › should allow changing the display style
- stock-filter.block\_theme.spec.ts:40 › Filter by Stock Block › should allow changing the block's title
- All Product Collection E2E tests.

cc @woocommerce/woo-fse

**Comment by senadir** [comment]
@dinhtungdu @sunyatasattva are you going to revisit this? I'm going over Cart/Checkout blocks to confirm them and make sure they're working.

**Comment by senadir** [comment]
Done testing and updating Cart/Checkout/Mini Cart/Order confirmation blocks/templates, will look at any possible broken tests next.

**Comment by dinhtungdu** [comment]
@senadir I looked at the All Products block today, the migration won't be smooth for that block as it isn't a functional component, so we can't use `useBlockProps` hook. I notice another issue with that block: in the editor, we can't select/focus that block from canvas anymore. I'm investigating more and will keep you posted.

**Comment by senadir** [comment]
Running Gutenberg seems to always force the iframe canvas and future versions of WP would probably do the same, so we might be able to get other blocks to V3 but still keep it in V1.

**Comment by samueljseay** [comment]
Migrating that block to functional components sounds like a good cooldown task for us, i'll add it to the backlog

**Comment by dinhtungdu** [comment]
> Running Gutenberg seems to always force the iframe canvas and future versions of WP would probably do the same, so we might be able to get other blocks to V3 but still keep it in V1.

@senadir I can confirm this. With this, we can upgrade blocks gradually instead of upgrading everything all at once, which is hard for reviewing, testing, and debugging. What do you think about dividing this PR into smaller ones, like one PR for cart and checkout, one for filters, one for product collection... If you agree:
- I will update the issue with a task list.
- Let's close this PR and use it as the reference.
- Look like it's all set for Rubik maintained block, right? Can you pick up related changes and create new PR(s) for Cart/Checkout/Mini Cart/Order confirmation blocks/templates?
- I will create PRs for blocks maintained by Store Editing.

**Comment by senadir** [comment]
Hey @dinhtungdu, we can do that, but would CI tests run with iframe or without it? I'm not sure if they're all running latest GB.

**Comment by senadir** [comment]
Spoke to @ellatrix and it's not super clear yet if WP 6.7 would ship with iframe always enabled or not, but we can assume that would be the case and we can break the PR.

**Comment by dinhtungdu** [comment]
@senadir I'm thinking about creating a custom CI job for this PR to run E2E tests in a custom environment that forces the iframe in the Post Editor. At the time of writing, WordPress nightly doesn't force the iframe when there is a non-v3 block registered. So our viable option is installing Gutenberg. Currently, we don't have blocks E2E tests run against Gutenberg nightly, our work here on a custom test environment isn't necessarily a one-off solution but can be turned into an optional CI job. What do you think?

**Comment by dinhtungdu** [comment]
Update: I'm updating E2E tests, we may not need to split this PR anymore.
Edit: All block E2E tests have been fixed.

**Review by WunderBart — COMMENTED** [review]
I've reviewed it from the perspective of the E2E tests, and it looks good! Nice work! 🚀

File: packages/js/product-editor/src/blocks/generic/checkbox/block.json [hunk 1]
```diff
@@ -38,5 +42,7 @@
 		"__experimentalToolbar": false
 	},
 	"editorStyle": "file:./editor.css",
-	"usesContext": [ "postType" ]
-}
+	"usesContext": [
+		"postType"
+	]
+}
```
Comment (Aljullu) [inline packages/js/product-editor/src/blocks/generic/checkbox/block.json:48]: Is it intentional that the ending blank line has been removed in several `block.json` files?

**Review by Aljullu — COMMENTED** [review]
I did some smoke testing and couldn't find anything broken. Code changes also look good. Fantastic job!

I left one minor comment about code style, but that's all.

> With a plugin/theme containing a v1 or v2 block.

Do you have a recommendation for this? I used [CoBlocks](https://wordpress.org/plugins/coblocks/), but it might be a good idea to provide an example in the testing steps. :thinking:

File: packages/js/product-editor/src/blocks/generic/checkbox/block.json (hunk 1 above)
Comment (dinhtungdu) [reply packages/js/product-editor/src/blocks/generic/checkbox/block.json:48]: I think it's auto formatted by editor. The formatting config is different from mine btw. @senadir can you confirm?

We don't have a configuration for JSON, which I think we should add one, using prettier.

**Comment by dinhtungdu** [comment]
> Do you have a recommendation for this? I used [CoBlocks](https://wordpress.org/plugins/coblocks/), but it might be a good idea to provide an example in the testing steps. 🤔

Good call! I've just created an example plugin for this PR and updated the testing description. Thanks!

**Comment by dinhtungdu** [comment]
> Do you have a recommendation for this? I used [CoBlocks](https://wordpress.org/plugins/coblocks/), but it might be a good idea to provide an example in the testing steps. 🤔

@Aljullu Good call! I've just created an example plugin for this PR and updated the testing description. Thanks!

**Review by kmanijak — COMMENTED** [review]
Smoke tested this with and without blocks on a lower API version. Didn't encounter any problems with blocks.

Amazing job!

File: packages/js/product-editor/src/blocks/generic/checkbox/block.json (hunk 1 above)
Comment (senadir) [reply packages/js/product-editor/src/blocks/generic/checkbox/block.json:48]: Yeah that was auto formatted.

**Review by mikejolley — APPROVED** [review]
@senadir I checked Cart, Checkout, Mini Cart. I didn't notice regressions and I tested editable fields which also appeared to work fine. 

I scanned over the code and nothing stood out. I did see tests have been updated—would it be possible do you think to have a suite running tests with that v2 API test plugin active so we can ensure things work in both iframe and non iframe mode?

I will approve this my side but other teams should test other block types.

**Comment by dinhtungdu** [comment]
> would it be possible do you think to have a suite running tests with that v2 API test plugin active so we can ensure things work in both iframe and non iframe mode?

This will require updating all updated tests to fallback to `page` object since `editor.canvas` [doesn't do that anymore](https://github.com/WordPress/gutenberg/pull/54911). @WunderBart do you have any other idea how we can achieve this?

**Comment by WunderBart** [comment]
> > would it be possible do you think to have a suite running tests with that v2 API test plugin active so we can ensure things work in both iframe and non iframe mode?
> 
> This will require updating all updated tests to fallback to `page` object since `editor.canvas` [doesn't do that anymore](https://github.com/WordPress/gutenberg/pull/54911). @WunderBart do you have any other idea how we can achieve this?

This is a somewhat complex issue because we’re utilizing the same utilities (e.g., the `editor.canvas` locator) in both the Post Editor and Site Editor contexts. The difference is that the Post Editor uses an iframe only for v3 blocks, while the Site Editor always uses it. Before diving into the details, let’s clarify the requirements: When running for v2 blocks, would that apply to specific tests, or would it involve all the Editor tests? In other words, do we want the option to run only certain tests with v2 blocks (dynamic approach), or will we always be executing the entire E2E codebase with v2 OR v3 blocks (static approach)?

**Comment by senadir** [comment]
I don't see a huge value we would gain by adding tests for V2, we can at best, add a test to ensure we're not regressing back to V2? Otherwise, as time goes forward, Gutenberg will do more effort to always load up V3.

**Comment by senadir** [comment]
Added a test that ensures the post editor is loaded in iframe mode.

File: plugins/woocommerce-blocks/tests/e2e/tests/basic.block_theme.spec.ts [hunk 2]
```diff
@@ -13,6 +13,12 @@ test.describe( 'Basic role-based functionality tests', () => {
 				page.getByRole( 'heading', { name: 'Dashboard' } )
 			).toHaveText( 'Dashboard' );
 		} );
+
+		test( 'Load post editor in iframe mode', async ( { page } ) => {
+			await page.goto( '/wp-admin/post-new.php' );
+
+			// Check that iframe with title "Editor canvas" is visible.
+			await expect( page.getByTitle( 'Editor canvas' ) ).toBeVisible();
```
Comment (senadir) [inline plugins/woocommerce-blocks/tests/e2e/tests/basic.block_theme.spec.ts]: Wasn't sure what's the best way to go about this.

**Comment by dinhtungdu** [comment]
> Before diving into the details, let’s clarify the requirements: When running for v2 blocks, would that apply to specific tests, or would it involve all the Editor tests? In other words, do we want the option to run only certain tests with v2 blocks (dynamic approach), or will we always be executing the entire E2E codebase with v2 OR v3 blocks (static approach)?

@WunderBart for your question, my answer is the latter, a static approach. But I agree with @senadir that it seems to not add much value as Gutenberg will always load the iframe in the future.

> add a test to ensure we're not regressing back to V2?

@senadir with https://github.com/WordPress/gutenberg/pull/54911 all tests in the Post editor will fail if the iframe isn't loaded.

File: plugins/woocommerce-blocks/tests/e2e/tests/basic.block_theme.spec.ts (hunk 2 above)
Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/tests/basic.block_theme.spec.ts]: My suggestion would be to use the [`frameLocator.owner`](https://playwright.dev/docs/api/class-framelocator#frame-locator-owner) API here:

```suggestion

		test( 'Load iframed post editor', async ( { admin, editor } ) => {
			await admin.createNewPost();
			await expect( editor.canvas.owner() ).toBeVisible();
```

**Pull request merged**